File: Data Structures & Algorithms/same-binary-tree/submission-5.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isSameTree(p: TreeNode?, q: TreeNode?): Boolean {
        val stack = ArrayDeque<Pair<TreeNode?, TreeNode?>>(listOf(Pair(p, q)))

        while (stack.isNotEmpty()) {
            val (first, second) = stack.removeLast()

            if (first == null && second == null) continue
            if (first == null || second == null || first?.`val` != second.`val`) return false

            stack.add(Pair(first?.left, second?.left))
            stack.add(Pair(first?.right, second?.right))
        }
        
        return true
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        left = self.getHeight(root.left)
        right = self.getHeight(root.right)

        if abs(left - right) > 1:
            return False

        return self.isBalanced(root.left) and self.isBalanced(root.right)

    def getHeight(self, root: Optional[TreeNode], depth = 0) -> int:
        if not root:
            return depth

        return max(self.getHeight(root.left, depth + 1), self.getHeight(root.right, depth + 1))